File: clients/python/model/player.py
from model.vehicle import Vehicle
from stream_wrapper import StreamWrapper
from typing import List

class Player:
    """TODO - Document"""

    __slots__ = ("index","score","vehicles",)

    index: int
    score: int
    vehicles: List[Vehicle]

    def __init__(self, index: int, score: int, vehicles: List[Vehicle]):
        self.index = index
        """TODO - Document"""
        self.score = score
        """TODO - Document"""
        self.vehicles = vehicles
        """TODO - Document"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "Player":
        """Read Player from input stream
        """
        index = stream.read_int()
        score = stream.read_long()
        vehicles = []
        for _ in range(stream.read_int()):
            vehicles_element = Vehicle.read_from(stream)
            vehicles.append(vehicles_element)
        return Player(index, score, vehicles)
    
    def write_to(self, stream: StreamWrapper):
        """Write Player to output stream
        """
        stream.write_int(self.index)
        stream.write_long(self.score)
        stream.write_int(len(self.vehicles))
        for element in self.vehicles:
            element.write_to(stream)
    
    def __repr__(self):
        return "Player(" + \
            repr(self.index) + \
            ", " + \
            repr(self.score) + \
            ", " + \
            repr(self.vehicles) + \
            ")"

File: clients/go/model/player.go
package model

import "fmt"
import "io"
import . "komus24/stream"

// TODO - Document
type Player struct {
    // TODO - Document
    Index int32
    // TODO - Document
    Score int64
    // TODO - Document
    Vehicles []Vehicle
}

func NewPlayer(index int32, score int64, vehicles []Vehicle) Player {
    return Player {
        Index: index,
        Score: score,
        Vehicles: vehicles,
    }
}

// Read Player from reader
func ReadPlayer(reader io.Reader) Player {
    var index int32
    index = ReadInt32(reader)
    var score int64
    score = ReadInt64(reader)
    var vehicles []Vehicle
    vehicles = make([]Vehicle, ReadInt32(reader))
    for vehiclesIndex := range vehicles {
        var vehiclesElement Vehicle
        vehiclesElement = ReadVehicle(reader)
        vehicles[vehiclesIndex] = vehiclesElement
    }
    return Player {
        Index: index,
        Score: score,
        Vehicles: vehicles,
    }
}

// Write Player to writer
func (player Player) Write(writer io.Writer) {
    index := player.Index
    WriteInt32(writer, index)
    score := player.Score
    WriteInt64(writer, score)
    vehicles := player.Vehicles
    WriteInt32(writer, int32(len(vehicles)))
    for _, vehiclesElement := range vehicles {
        vehiclesElement.Write(writer)
    }
}

// Get string representation of Player
func (player Player) String() string {
    stringResult := "{ "
    stringResult += "Index: "
    index := player.Index
    stringResult += fmt.Sprint(index)
    stringResult += ", "
    stringResult += "Score: "
    score := player.Score
    stringResult += fmt.Sprint(score)
    stringResult += ", "
    stringResult += "Vehicles: "
    vehicles := player.Vehicles
    stringResult += "[ "
    for vehiclesIndex, vehiclesElement := range vehicles {
        if vehiclesIndex != 0 {
            stringResult += ", "
        }
        stringResult += vehiclesElement.String()
    }
    stringResult += " ]"
    stringResult += " }"
    return stringResult
}